File: Robotics Kinematics and Dynamics Course/Chapter4/Lesson3/Chapter4_Lesson3.py
import numpy as np

def rot2(theta):
    """2D rotation matrix."""
    c = np.cos(theta)
    s = np.sin(theta)
    return np.array([[c, -s],
                     [s,  c]])

def fk_planar_2R(q, L):
    """
    Open-chain forward kinematics for a planar 2R arm.
    q: array-like [theta1, theta2]
    L: array-like [L1, L2] link lengths
    Returns end-effector position in R^2.
    """
    theta1, theta2 = q
    L1, L2 = L

    p1 = rot2(theta1) @ np.array([L1, 0.0])
    p2 = p1 + rot2(theta1 + theta2) @ np.array([L2, 0.0])
    return p2

def fourbar_constraint(theta1, theta3, L0, L1, L2, L3):
    """
    Loop-closure constraint Phi(theta1, theta3) for the four-bar example.
    Returns scalar Phi; feasible configurations satisfy Phi = 0.
    """
    pB = np.array([L1 * np.cos(theta1),
                   L1 * np.sin(theta1)])
    pC = np.array([L0 + L3 * np.cos(theta3),
                   0.0 + L3 * np.sin(theta3)])
    Phi = np.dot(pB - pC, pB - pC) - L2**2
    return Phi

def solve_fourbar_theta3(theta1, L0, L1, L2, L3,
                         theta3_init=0.0,
                         tol=1e-10,
                         max_iter=50):
    """
    Solve Phi(theta1, theta3) = 0 for theta3 using Newton iteration.
    For robustness in practice, bracketed solvers or multiple initial
    guesses are recommended.
    """
    theta3 = float(theta3_init)
    for k in range(max_iter):
        Phi = fourbar_constraint(theta1, theta3, L0, L1, L2, L3)

        # Numerical derivative dPhi/dtheta3 via symmetric difference
        h = 1e-6
        Phi_p = fourbar_constraint(theta1, theta3 + h, L0, L1, L2, L3)
        Phi_m = fourbar_constraint(theta1, theta3 - h, L0, L1, L2, L3)
        dPhi = (Phi_p - Phi_m) / (2.0 * h)

        if abs(dPhi) < 1e-12:
            break  # avoid division by zero

        step = Phi / dPhi
        theta3 -= step

        if abs(step) < tol:
            break
    return theta3

if __name__ == "__main__":
    # Open chain: 2R arm
    L = [0.5, 0.4]
    q = [0.5, -0.3]
    p_ee = fk_planar_2R(q, L)
    print("Open-chain end-effector position:", p_ee)

    # Closed chain: four-bar linkage
    L0, L1, L2, L3 = 0.8, 0.5, 0.7, 0.4
    theta1 = 0.4
    theta3 = solve_fourbar_theta3(theta1, L0, L1, L2, L3, theta3_init=0.2)
    print("Closed-chain solution theta3:", theta3)
    print("Constraint residual:",
          fourbar_constraint(theta1, theta3, L0, L1, L2, L3))
      

File: Tutorials/Codes/Robotics_Kinematics_and_Dynamics_Course/Chapter4/Lesson3/Chapter4_Lesson3.java
public class OpenClosedKinematics {

    public static double[] rot2TimesVec(double theta, double[] v) {
        double c = Math.cos(theta);
        double s = Math.sin(theta);
        return new double[] {
            c * v[0] - s * v[1],
            s * v[0] + c * v[1]
        };
    }

    public static double[] fkPlanar2R(double theta1, double theta2,
                                      double L1, double L2) {
        double[] e1 = new double[] {L1, 0.0};
        double[] e2 = new double[] {L2, 0.0};

        double[] p1 = rot2TimesVec(theta1, e1);
        double[] e2rot = rot2TimesVec(theta1 + theta2, e2);

        double[] p2 = new double[2];
        p2[0] = p1[0] + e2rot[0];
        p2[1] = p1[1] + e2rot[1];
        return p2;
    }

    public static double fourbarConstraint(double theta1, double theta3,
                                           double L0, double L1,
                                           double L2, double L3) {
        double xB = L1 * Math.cos(theta1);
        double yB = L1 * Math.sin(theta1);
        double xC = L0 + L3 * Math.cos(theta3);
        double yC =        L3 * Math.sin(theta3);
        double dx = xB - xC;
        double dy = yB - yC;
        return dx*dx + dy*dy - L2*L2;
    }

    public static double solveFourbarTheta3(double theta1,
                                            double L0, double L1,
                                            double L2, double L3,
                                            double theta3Init,
                                            double tol, int maxIter) {
        double theta3 = theta3Init;
        for (int k = 0; k < maxIter; ++k) {
            double Phi = fourbarConstraint(theta1, theta3, L0, L1, L2, L3);
            double h = 1e-6;
            double PhiP = fourbarConstraint(theta1, theta3 + h, L0, L1, L2, L3);
            double PhiM = fourbarConstraint(theta1, theta3 - h, L0, L1, L2, L3);
            double dPhi = (PhiP - PhiM) / (2.0 * h);
            if (Math.abs(dPhi) < 1e-14) break;
            double step = Phi / dPhi;
            theta3 -= step;
            if (Math.abs(step) < tol) break;
        }
        return theta3;
    }

    public static void main(String[] args) {
        double theta1 = 0.5, theta2 = -0.3;
        double L1 = 0.5, L2 = 0.4;
        double[] p = fkPlanar2R(theta1, theta2, L1, L2);
        System.out.println("Open-chain end-effector: (" +
                           p[0] + ", " + p[1] + ")");

        double L0 = 0.8, L3 = 0.4, Lmid = 0.7;
        double theta1In = 0.4;
        double theta3 = solveFourbarTheta3(theta1In, L0, L1, Lmid, L3,
                                           0.2, 1e-10, 50);
        System.out.println("Closed-chain theta3: " + theta3);
        System.out.println("Constraint residual: " +
            fourbarConstraint(theta1In, theta3, L0, L1, Lmid, L3));
    }
}
      
